Update for Battery Efficiency Fix Issue #11277

bobbycisneros opened the pull request:
This update has two objectives: 
1. Fix the battery efficiency OSD element.  With British units, the display reads "-----" whenever the speed exceeds 47 miles/hour.  The original logic scaled the gps speed in cm/sec by 100 then converted it to miles/hr.  In the process, whenever the GPS speed exceeds 47 miles/hr the return value goes invalid. The fix for this is to remove the scaling, whereby keeping the number from going invalid.  
2. During testing I noticed the battery efficiency readout seemed to have a lot of noise and was "jumpy".  To address this, I added a filter to the battery efficiency calculation.  This also improves readout at slow speed.

haslinghuis commented on src/main/osd/osd_elements.c:
```diff
@@ -178,6 +178,8 @@
 #define FULL_CIRCLE 360
 #define EFFICIENCY_MINIMUM_SPEED_CM_S 100
+#define EFFFILTERTC 0.1f
+static float efficiencyF = 0.0;
```
```suggestion
static float efficiencyF = 0.0f;
```

haslinghuis commented on src/main/osd/osd_elements.c:
```diff
@@ -1080,14 +1082,15 @@ static void osdElementGpsSpeed(osdElementParms_t *element)
 static void osdElementEfficiency(osdElementParms_t *element)
 {
-    int efficiency = 0;
+    int efficiency = 0.0;
```
```suggestion
    int efficiency = 0.0f;
```

bobbycisneros replied:
Good catch.  Shouldn't this be:

int efficiency = 0;

haslinghuis commented on src/main/osd/osd_elements.c:
```diff
@@ -1080,14 +1082,15 @@ static void osdElementGpsSpeed(osdElementParms_t *element)
+        if (speedX1 > 0) {
             const int mAmperage = getAmperage() * 10; // Current in mA
-            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
+            efficiencyF = efficiencyF + EFFFILTERTC * (mAmperage / speedX1 - efficiencyF); 
```
```suggestion
            efficiencyF += EFFFILTERTC * (mAmperage / speedX1 - efficiencyF); 
```

bobbycisneros commented:
Good recommendation!  I’ll get it done. 

> On Jan 28, 2022, at 10:56 PM, haslinghuis ***@***.***> wrote:
> 
> 
> @haslinghuis commented on this pull request.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/11361#discussion_r794999532>:
> 
> >              const int mAmperage = getAmperage() * 10; // Current in mA
> -            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
> +            efficiencyF = efficiencyF + EFFFILTERTC * (mAmperage / speedX1 - efficiencyF); 
> ⬇️ Suggested change
> -            efficiencyF = efficiencyF + EFFFILTERTC * (mAmperage / speedX1 - efficiencyF); 
> +            efficiencyF += EFFFILTERTC * (mAmperage / speedX1 - efficiencyF); 
> —
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/11361#pullrequestreview-866845381>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3CGNKNZBNYJFBMZV3LUYNXPDANCNFSM5NB6MHKA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you authored the thread.
>

haslinghuis commented on src/main/osd/osd_elements.c:
```diff
@@ -1080,14 +1082,15 @@ static void osdElementGpsSpeed(osdElementParms_t *element)
-            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
+            efficiencyF += EFFFILTERTC * (mAmperage / speedX1 - efficiencyF); 
         }
+        efficiency = (int)efficiencyF;
```
```suggestion
        efficiency = (int)Math.Round(efficiencyF);
```

bobbycisneros replied:
I couldn't find the Math.Round(xxx) function.  How about this:

efficiency = (int)round(efficiencyF);

?

KarateBrot replied:
@haslinghuis we have `int lrintf(float)` for that in the firmware

daleckystepan commented on src/main/osd/osd_elements.c:
```diff
@@ -1082,12 +1084,13 @@ static void osdElementEfficiency(osdElementParms_t *element)
-            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
+            efficiencyF += EFFFILTERTC * (mAmperage / speedX1 - efficiencyF); 
         }
+        efficiency = (int)round(efficiencyF);
```
Please use `lrintf` here. `round` uses doubles and we have to avoid it.

bobbycisneros replied:
Fixed!

daleckystepan commented:
@bobbycisneros Can you please edit the first post and describe what you are doing, how it works and why it is better than current approach? Thank you. It is hard to review without that.

bobbycisneros commented:
Update the description. As requested. Thanks. 

Sent from my iPhone

> On Feb 2, 2022, at 12:46 PM, Štěpán Dalecký ***@***.***> wrote:
> 
> ﻿
> Assigned #11361 to @bobbycisneros.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.

daleckystepan commented:
Thank you got it. `speedX1` can be just `speed`.

`mAmperage` and `speedX1` are integers. Be aware of integer division here.

Filter you have implemented is PT1 we already have in floats (which is also desired here). Check `filter.c` and `pt1FilterGain`, `pt1FilterInit` and `pt1FilterApply`.

bobbycisneros commented:
Thanks for the recommendations.  You and your team are rock star programmers.  I only do this once in a while since becoming a program manager for one of our products.  I implemented the recommended changes but want to test them before I publish them.  It will be a couple days since we recently had snow in our area.  

> On Feb 3, 2022, at 1:52 PM, Štěpán Dalecký ***@***.***> wrote:
> 
> 
> Thank you got it. speedX1 can be just speed.
> 
> mAmperage and speedX1 are integers. Be aware of integer division here.
> 
> Filter you have implemented is PT1 we already have in floats (which is also desired here). Check filter.c and pt1FilterGain, pt1FilterInit and pt1FilterApply.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/11361#issuecomment-1029345063>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3GMJSQJC3FVR7BJ3OTUZLMI7ANCNFSM5NB6MHKA>.
> You are receiving this because you were mentioned.
>

bobbycisneros commented:
Thanks Jan.  I already implemented this change.  

> On Feb 5, 2022, at 3:30 PM, Jan Post ***@***.***> wrote:
> 
> 
> @KarateBrot commented on this pull request.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/11361#discussion_r800101955>:
> 
> >          }
> +        efficiency = (int)efficiencyF;
> @haslinghuis <https://github.com/haslinghuis> we have int lrintf(float) for that in the firmware
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/11361#discussion_r800101955>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3HD2DQOLLSTSNPJ5LDUZWJHTANCNFSM5NB6MHKA>.
> You are receiving this because you were mentioned.
>

bobbycisneros commented:
Recommended changes were implemented and it worked in my test flight!  Checkout 95ed447 <https://github.com/betaflight/betaflight/pull/11361/commits/95ed44781285e14007ecc5208a210675fa912444>.  Let me know if there is anything else.  Thanks!

> On Feb 4, 2022, at 3:54 PM, Robert Cisneros ***@***.***> wrote:
> 
> Thanks for the recommendations.  You and your team are rock star programmers.  I only do this once in a while since becoming a program manager for one of our products.  I implemented the recommended changes but want to test them before I publish them.  It will be a couple days since we recently had snow in our area.  
> 
>> On Feb 3, 2022, at 1:52 PM, Štěpán Dalecký ***@***.*** ***@***.***>> wrote:
>> 
>> 
>> Thank you got it. speedX1 can be just speed.
>> 
>> mAmperage and speedX1 are integers. Be aware of integer division here.
>> 
>> Filter you have implemented is PT1 we already have in floats (which is also desired here). Check filter.c and pt1FilterGain, pt1FilterInit and pt1FilterApply.
>> 
>> —
>> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/11361#issuecomment-1029345063>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3GMJSQJC3FVR7BJ3OTUZLMI7ANCNFSM5NB6MHKA>.
>> You are receiving this because you were mentioned.
>> 
>

bobbycisneros commented:
Does anybody in this conversation know why my Linux builds are failing?  
Everything builds fine on MacOS.

KarateBrot commented:
```
linking ../../obj/test/link_quality_unittest/link_quality_unittest 
/usr/bin/ld: ../../obj/test/link_quality_unittest/osd/osd_elements.c.o: in function `osdElementsInit':
/home/vsts/work/1/s/src/test/../main/osd/osd_elements.c:1867: undefined reference to `pt1FilterUpdateCutoff'
```

You need to update the unit tests in osd_elements.c

daleckystepan commented on src/main/osd/osd_elements.c:
```diff
@@ -1858,6 +1863,8 @@ void osdElementsInit(bool backgroundLayerFlag)
 {
     backgroundLayerSupported = backgroundLayerFlag;
     activeOsdElementCount = 0;
+    pt1FilterInit(&batteryEfficiencyFilt, 0.0f);
```
`0.0f` is not good `gain`.  `gain` should be computed using `pt1FilterGain` based on

- `dT` sampling period in seconds (float number).
- `f_cut` - cutoff frequency in Hz


You probably don't need to call `pt1FilterUpdateCutoff` as it is the same `k` number as gain in Init.

KarateBrot replied:
```suggestion
    pt1FilterInit(&batteryEfficiencyFilt, pt1FilterGain(dT, EFFICIENCY_CUTOFF_HZ));
```
You need to replace `dT` with the looptime of the batt efficiency calculation.

KarateBrot commented on src/main/osd/osd_elements.c:
```diff
@@ -178,6 +179,9 @@
 
 #define FULL_CIRCLE 360
 #define EFFICIENCY_MINIMUM_SPEED_CM_S 100
+#define EFFFILTERTC 0.1f
```
```suggestion
#define EFFICIENCY_CUTOFF_HZ 10
```
Replace the value 10 with the cutoff frequency you want to get.

KarateBrot commented on src/main/osd/osd_elements.c:
```diff
@@ -1858,6 +1863,8 @@ void osdElementsInit(bool backgroundLayerFlag)
     backgroundLayerSupported = backgroundLayerFlag;
     activeOsdElementCount = 0;
+    pt1FilterInit(&batteryEfficiencyFilt, 0.0f);
+    pt1FilterUpdateCutoff(&batteryEfficiencyFilt, EFFFILTERTC);
```
```suggestion
```

KarateBrot reviewed (commented):
I explicitly typed in code what @daleckystepan already mentioned in words

KarateBrot commented on src/main/osd/osd_elements.c:
```diff
@@ -1858,6 +1864,7 @@ void osdElementsInit(bool backgroundLayerFlag)
 {
     backgroundLayerSupported = backgroundLayerFlag;
     activeOsdElementCount = 0;
+    pt1FilterInit(&batteryEfficiencyFilt, pt1FilterGain(EFFICIENCY_CUTOFF_HZ, 1.0f/50.0f));
```
In `pt1FilterGain()`: 
The first argument needs to be the looptime of the `osdElementEfficiency()` function and `EFFICIENCY_CUTOFF_HZ` needs to  be the second argument.

bobbycisneros replied:
I corrected the dt.  However the arguments are in the right order.

float pt1FilterGain(float f_cut, float dT)

KarateBrot replied:
Yeah sorry my bad. I corrected my comment.

daleckystepan commented on src/main/osd/osd_elements.c:
```diff
@@ -1082,12 +1087,13 @@ static void osdElementEfficiency(osdElementParms_t *element)
+        const int speed = osdGetSpeedToSelectedUnit(gpsSol.groundSpeed); 
         
-        if (speedX100 > 0) {
+        if (speed > 0) {
```
Do we really need this condition and static variable `efficiencyF`?
If speed is really == 0 then you will see only the last value not the zero. Also calling `pt1FilterApply` with uneven rate is not good for the filter.

Cutoff `#define EFFICIENCY_CUTOFF_HZ 0.212206591f` is also pretty small and weird number.


We can change `mAmperage` and `speed` type to `float` to avoid explicit type conversion.

bobbycisneros replied:
I changed the cutoff to 0.2, low but adequate.  The filter is just trying to smooth the efficiency OSD display so it's easier to read, it really doesn't matter if it misses a few frames when speed == 0.  Besides we need the divide by zero protection.

bobbycisneros replied:
corrected

daleckystepan commented on src/main/osd/osd_elements.c:
```diff
@@ -1858,6 +1864,7 @@ void osdElementsInit(bool backgroundLayerFlag)
 {
     backgroundLayerSupported = backgroundLayerFlag;
     activeOsdElementCount = 0;
+    pt1FilterInit(&batteryEfficiencyFilt, pt1FilterGain(EFFICIENCY_CUTOFF_HZ, 1.0f/OSD_FRAMERATE_DEFAULT_HZ));
```
I would prefer `1.0f/osdConfig()->framerate_hz` here as `dT`.

bobbycisneros replied:
Updated as requested.

bobbycisneros replied:
Corrected

haslinghuis commented on src/main/osd/osd_elements.c:
```diff
@@ -1082,12 +1086,13 @@ static void osdElementEfficiency(osdElementParms_t *element)
-            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
+        if (speed > 0) {
+            const float mAmperage = getAmperage() * 10; // Current in mA
+            pt1FilterApply(&batteryEfficiencyFilt, (mAmperage/speed));
```
```suggestion
            pt1FilterApply(&batteryEfficiencyFilt, (mAmperage / speed));
```

bobbycisneros replied:
Corrected

haslinghuis commented on src/main/osd/osd_elements.c:
```diff
@@ -1858,6 +1863,7 @@ void osdElementsInit(bool backgroundLayerFlag)
 {
     backgroundLayerSupported = backgroundLayerFlag;
     activeOsdElementCount = 0;
+    pt1FilterInit(&batteryEfficiencyFilt, pt1FilterGain(EFFICIENCY_CUTOFF_HZ, 1.0f/osdConfig()->framerate_hz));
```
```suggestion
    pt1FilterInit(&batteryEfficiencyFilt, pt1FilterGain(EFFICIENCY_CUTOFF_HZ, 1.0f / osdConfig()->framerate_hz));
```

bobbycisneros replied:
done

haslinghuis reviewed (changes requested):
For better readability it's better to have space around operators. Otherwise LGTM.

bobbycisneros commented:
Squash complete.
Finally done!  Thanks

> On Feb 12, 2022, at 11:20 AM, Jan Post ***@***.***> wrote:
> 
> 
> ...and please squash the commits into 1 single commit :)
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/11361#issuecomment-1037325914>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3CQM55E7SYJ6ME2SB3U22JHLANCNFSM5NB6MHKA>.
> You are receiving this because you were mentioned.
>

KarateBrot commented on src/main/osd/osd_elements.c:
```diff
@@ -1082,12 +1086,12 @@ static void osdElementEfficiency(osdElementParms_t *element)
-            const int mAmperage = getAmperage() * 10; // Current in mA
-            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
+        const float speed = (float)osdGetSpeedToSelectedUnit(gpsSol.groundSpeed);
+        if (speed > 0.0f) {
```
Wait a sec... 

`if (gpsSol.groundSpeed >= EFFICIENCY_MINIMUM_SPEED_CM_S)`

is already tested in line 1088, doesn't that mean that `float speed` is already guaranteed to be greater than `0.0f` anyways? In other words: can we remove this conditional in line 1090?

bobbycisneros replied:
Yes it does appear to be redundant. I was modeling it after the original code which Checked to be sure speedX100 was greater than zero. I'll remove the extra condition. Do I need to squash the two commits?

KarateBrot replied:
Yeah please. But you can also do
```
git reset --soft HEAD~1
```
then remove the conditional and then
```
git add -a
git commit -m "Update for Battery Efficiency Fix Issue #11277"
git push --force
```

haslinghuis replied:
Or 
```
git commit -a --amend --no-edit
git push origin +OSDBattEfficiencyFix
```

bobbycisneros commented:
Final update made as requested.

> On Feb 12, 2022, at 6:27 PM, Jan Post ***@***.***> wrote:
> 
> 
> @KarateBrot commented on this pull request.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/11361#discussion_r805219733>:
> 
> > @@ -1082,12 +1086,12 @@ static void osdElementEfficiency(osdElementParms_t *element)
>  {
>      int efficiency = 0;
>      if (sensors(SENSOR_GPS) && ARMING_FLAG(ARMED) && STATE(GPS_FIX) && gpsSol.groundSpeed >= EFFICIENCY_MINIMUM_SPEED_CM_S) {
> -        const int speedX100 = osdGetSpeedToSelectedUnit(gpsSol.groundSpeed * 100); // speed * 100 for improved resolution at slow speeds
> -        
> -        if (speedX100 > 0) {
> -            const int mAmperage = getAmperage() * 10; // Current in mA
> -            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
> +        const float speed = (float)osdGetSpeedToSelectedUnit(gpsSol.groundSpeed);
> +        if (speed > 0.0f) {
> Yeah please. But you can also do
> 
> git reset --soft HEAD~1
> then remove the conditional and then
> 
> git add -a
> git commit -m "Update for Battery Efficiency Fix Issue #11277"
> git push --force
> —
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/11361#discussion_r805219733>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3CXCZLOCUN2NXGNOB3U233GLANCNFSM5NB6MHKA>.
> You are receiving this because you were mentioned.
>

daleckystepan commented on src/main/osd/osd_elements.c:
```diff
@@ -1082,12 +1086,10 @@ static void osdElementEfficiency(osdElementParms_t *element)
-        }
+        const float speed = (float)osdGetSpeedToSelectedUnit(gpsSol.groundSpeed);
+        const float mAmperage = (float)getAmperage() * 10.f; // Current in mA
+        pt1FilterApply(&batteryEfficiencyFilt, (mAmperage / speed));
```
`efficiency = lrintf(pt1FilterApply(&batteryEfficiencyFilt, mAmperage / speed));`

Will be better than using internal structure like on the next line.

bobbycisneros commented:
Updated as requested and squashed in with the other commits.

> On Feb 13, 2022, at 3:48 AM, Štěpán Dalecký ***@***.***> wrote:
> 
> 
> @daleckystepan commented on this pull request.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/11361#discussion_r805308860>:
> 
> > @@ -1082,12 +1086,10 @@ static void osdElementEfficiency(osdElementParms_t *element)
>  {
>      int efficiency = 0;
>      if (sensors(SENSOR_GPS) && ARMING_FLAG(ARMED) && STATE(GPS_FIX) && gpsSol.groundSpeed >= EFFICIENCY_MINIMUM_SPEED_CM_S) {
> -        const int speedX100 = osdGetSpeedToSelectedUnit(gpsSol.groundSpeed * 100); // speed * 100 for improved resolution at slow speeds
> -        
> -        if (speedX100 > 0) {
> -            const int mAmperage = getAmperage() * 10; // Current in mA
> -            efficiency = mAmperage * 100 / speedX100; // mAmperage * 100 to cancel out speed * 100 from above
> -        }
> +        const float speed = (float)osdGetSpeedToSelectedUnit(gpsSol.groundSpeed);
> +        const float mAmperage = (float)getAmperage() * 10.f; // Current in mA
> +        pt1FilterApply(&batteryEfficiencyFilt, (mAmperage / speed));
> efficiency = lrintf(pt1FilterApply(&batteryEfficiencyFilt, mAmperage / speed));
> 
> Will be better than using internal structure like on the next line.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/11361#pullrequestreview-880959179>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3BL4GEVE52BMLZUAGLU2547XANCNFSM5NB6MHKA>.
> You are receiving this because you were mentioned.
>
